### PR TITLE
extend the identity proof manager to allow system wide key pairs

### PR DESCRIPTION
**Pull request by schiessle**
extend the identity proof manager to allow system wide key pairs

### CODE REVIEW

File: lib/private/Security/IdentityProof/Manager.php [hunk 1]
```diff
@@ -69,17 +76,17 @@ protected function generateKeyPair() {
 	 * Generate a key for $user
```
Comment (LukasReschke) [inline lib/private/Security/IdentityProof/Manager.php]: PHPDoc lies now :)

File: lib/private/Security/IdentityProof/Manager.php [hunk 2]
```diff
@@ -88,22 +95,42 @@ protected function generateKey(IUser $user) {
...
+	 *
+	 * @param IUser $user
+	 * @return Key
+	 */
+	public function getKey(IUser $user) {
+		return $this->retrieveKey($user->getUID());
+	}
+
+	/**
+	 * get instance wide public and private key
```
Comment (LukasReschke) [inline lib/private/Security/IdentityProof/Manager.php]: Uppercase the first letter here :)

**Review by LukasReschke — CHANGES_REQUESTED** [review]
PHPDoc needs a small chance otherwise I think this is okay but would like @rullzer's thoughts as well :)

**Comment by nickvergessen** [comment]
Looks good otherwise

**Comment by rullzer** [comment]
So... uhm... on the one hand sure...

So the collision chance is practically zero. I just want to point out that it still could happen. So if anything is encrypted with this and a collision does happen you will be out of luck. (And you'll not even know in the current state).

I don't know what you want to do it with. But if loss of this key due to collision is really an issue. Maybe also hash it?

/me will now go out of paranoid mode

**Comment by schiessle** [comment]
@rullzer what kind of collision do you mean? That there is a user with the same uid as the instance id? In this case I don't see how a hash would help, there you have the same risk that you have a uid which is the same as the hash of the instance id.

Even if this collision would happen, this would "only" mean that this one user would use the same key as the system. I don't see how this would resolve in data lose.

**Comment by rullzer** [comment]
Ah fair enough. Ok lets do it then.

**Comment by schiessle** [comment]
@LukasReschke I updated the documentation

**Comment by LukasReschke** [comment]
> @rullzer what kind of collision do you mean? That there is a user with the same uid as the instance id? In this case I don't see how a hash would help, there you have the same risk that you have a uid which is the same as the hash of the instance id.

Hash collisions are super unlikely and there is no known hash collision for sane hash mechanisms such as SHA-512. (and having a username that is multiple terrabytes or gigabytes long, this won't work with our DB scheme anyways :)) 

As in the future this here, could be extended in a way that a user may be able to export or import their private certificate or such having a collision here would be problematic. Especially as the instance id is known to the public and some instances have registrations apps or guest apps etc…

I'd prefer moving the systemkey in another folder thus or add a prefix to user keys starting with `user-` and a prefix to system keys starting with `system-` and as a fallback we read the old files. (or we even add a migration step, which shouldn't be super hurtful as I doubt that there are many instances with many of those keys yet as they are only generated on-demand)

**Comment by schiessle** [comment]
> Hash collisions are super unlikely and there is no known hash collision for sane hash mechanisms such as SHA-512.

It is not about hash collisions. It is about whether it is more likely that there will be a user with uid === instanceID or uid === hash(instanceId), I think there is no real difference. Both are extreme unlikely but of course could happen.

> I'd prefer moving the systemkey in another folder thus or add a prefix to user keys starting with user- and a prefix to system keys starting with system- and as a fallback we read the old files. 

I think just moving the systemkey to a subfolder wouldn't help much. There you also have the risk of a name collision. I will look into the possibility to add a prefix for each key.

**Comment by LukasReschke** [comment]
> It is not about hash collisions. It is about whether it is more likely that there will be a user with uid === instanceID or uid === hash(instanceId), I think there is no real difference. Both are extreme unlikely but of course could happen.

1. Have registration app installed
2. Have someone register with the instance id as username

The instance id is publicly known knowledge as it's the cookie name 🙈

**Comment by schiessle** [comment]
I added now a prefix `user-` and `system-` to avoid name collisions

**Review by rullzer — APPROVED** [review]

**Comment by schiessle** [comment]
@LukasReschke waiting for your +1 😉

**Review by LukasReschke — APPROVED** [review]

**Pull request merged**